Syntax lookup: Polyvar (#318)
rescript-association/rescript-lang.org: misc_docs/syntax/language_polyvar.mdx

---
id: "polyvar"
keywords: ["polymorphic", "variant", "polyvar"]

> name: "[ #value ]"

Reviewer: Maybe let's go just with `#value`, since it's the significant part

```suggestion
name: "#value"
```

> summary: "This is the `polymorphic variant` type declaration."

Reviewer: ```suggestion
summary: "This is a `polymorphic variant`."
```

category: "languageconstructs"
---

> [Polymorphic variants](/docs/manual/latest/polymorphic-variant) (or _poly variants_) are a specific kind of [variant](/docs/manual/latest/variant).

Reviewer: ```suggestion
[Polymorphic variants](/docs/manual/latest/polymorphic-variant) (or _poly variants_) are the structurally typed equivalent of [variants](/docs/manual/latest/variant).
```

> Their usage doesn't require any explicit type definition, and they not coupled to any specific module. The compiler will infer the type on demand, and compare poly variants by their value, instead of their type name.

Reviewer: ```suggestion
In comparison to nominally typed variants, their values don't require any explicit type definition, and are not coupled to any specific module. The compiler will infer the type on demand, and compare poly variants by their value, instead of their type name.
```

> However, using the `[ #value ]` syntax you can to define a _closed poly variant type_ with an exact set of values, known as constructors.

Reviewer: ```suggestion
However, you can also describe a set of polymorphic variants with a type declaration, which can be useful to explicitly annotate function parameters and values. The most common format is the _closed poly variant_ type definition that looks like this: `type x = [ #value ]` (Note the extra `[]` around the poly var constructors!).
```
